editoast/src/projects.rs

```diff
@@ -0,0 +1,186 @@
+use crate::error::Result;
+use crate::tables::osrd_infra_project;
+use crate::tables::osrd_infra_project::dsl;
+use chrono::NaiveDateTime;
+use diesel::result::Error as DieselError;
+use diesel::sql_types::{Array, Integer, Nullable, Text};
+use diesel::ExpressionMethods;
+use diesel::{delete, sql_query, update, PgConnection, QueryDsl, RunQueryDsl};
+use editoast_derive::EditoastError;
+use serde_json::{json, Map, Value};
+use thiserror::Error;
+
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Error, EditoastError)]
+#[editoast_error(base_id = "projects", context = "Self::context")]
+pub enum ProjectError {
+    /// Couldn't found the infra with the given id
+    #[error("Project '{0}', could not be found")]
+    #[editoast_error(status = 404)]
+    NotFound(i64),
+}
+
+impl ProjectError {
+    fn context(&self) -> Map<String, Value> {
+        match self {
+            ProjectError::NotFound(project_id) => json!({
+                "project_id": project_id,
+            })
+            .as_object()
+            .cloned()
+            .unwrap(),
+        }
+    }
+}
+
+#[derive(
+    Clone, QueryableByName, Queryable, Debug, Serialize, Deserialize, AsChangeset, Identifiable,
+)]
+#[diesel(table_name = osrd_infra_project)]
+pub struct Project {
+    pub id: i64,
+    pub name: String,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub creation_date: NaiveDateTime,
+    pub last_modification: NaiveDateTime,
+    pub tags: Option<Vec<String>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ProjectData {
+    pub name: String,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub tags: Option<Vec<String>>,
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct ProjectPatch {
+    pub name: Option<String>,
+    pub description: Option<String>,
+    pub objectives: Option<String>,
+    pub funders: Option<Vec<String>>,
+    pub budget: Option<i32>,
+    pub image: Option<Vec<u8>>,
+    pub tags: Option<Vec<String>>,
+}
+
+impl Project {
+    pub fn create(data: ProjectData, conn: &mut PgConnection) -> Result<Project> {
+        match sql_query(
+        "INSERT INTO osrd_infra_project (name, description, objectives, funders, budget, image, creation_date,last_modification, tags)
+         VALUES ($1, $2, $3, $4, $5, NULL,CURRENT_TIMESTAMP, CURRENT_TIMESTAMP,$6)
+         RETURNING *",
+    )
```

Comment:
Multiple things here. 

- Check out the diesel documentation https://diesel.rs/guides/all-about-inserts.html. In the **Insertable** section you can find an example that uses the `Insertable` derivative trait.
- Adapt `ProjectData` so you can directly insert an object of this type in the table. You can also rename it `ProjectCreateForm`.
- Stop using raw SQL. Using diesel dsl will avoid SQL errors.